File: June-LeetCoding-Challenge-2020/Week2/392-is-subsequence.cpp
// Time Complexity: O(max(s.length(), t.length()))
// Space Complexity: O(s.length())
// Runtime: 4 ms, faster than 85.40%
// Memory Usage: 6.5 MB, less than 84.45%
class Solution {
public:

 //s = "abc", t = "ahbgdc"
    bool isSubsequence(string s, string t) {
        
        if(s.length() > t.length())
            return false;
        
        if(!s.length())
            return true;
        
        if(!t.length())
            return false;
        
        queue<char> q;
        
        for(int i = 0; i < s.length(); i++){
            q.push(s[i]);
        }
        
        for(int i = 0; i < t.length(); i++){
            if(t[i] == q.front()){
                q.pop();
            }
        }
        
        return q.empty();
    }
};

// Time Complexity: O(t)
// Space Complexity: O(1)
// Runtime: 4 ms, faster than 84.88%
// Memory Usage: 6.3 MB, less than 95.10%
class Solution {
public:

    bool isSubsequence(string s, string t) {
        
        if(s.length() > t.length())
            return false;
        
        if(!s.length())
            return true;
        
        if(!t.length())
            return false;
        
        int pointer = 0;

        for(char comp : t){
            if(s[pointer] == comp){
                pointer++;
            }
        }
        
        return pointer == s.length();
    }
};

File: June-LeetCoding-Challenge-2020/Week2/392-is-subsequence.py
class Solution:
    # Sol 1. Using Pointer
    def isSubsequence(self, s: str, t: str) -> bool:
        if s == "":
            return True
        
        ptr = 0
        for c in t:
            if s[ptr] == c:
                ptr += 1
                if ptr == len(s):
                    return True
        return False
    # TC: O(n), where n is length of t, best case n is the length of s
    # SC: O(1)
    # Runtime: 40 ms, faster than 60.80%
    # Memory Usage: 13.7 MB, less than 98.08%

    # Sol 2. Brute Force, Not working
    def isSubsequence(self, s:str, t:str) -> bool:
        if len(s)==0:
            return True
        elif len(t)==0:
            return False

        s_list = [char for char in s]
        t_list = [char for char in t]
        ret_list = []
        ret = False
        # not working because it compares un-ordered
        for s_c in s_list:
            for t_c in t_list:
                if s_c == t_c:
                    ret_list.append(t_c)
        print(s_list)
        print(t_list)
        print(ret_list)

        return ret_list == s_list